### PR TITLE
Validate ml2

### PR DESCRIPTION
**Pull request by toabctl**

### CODE REVIEW

**Comment by rhafer** [comment]
LGTM

File: crowbar_framework/app/models/neutron_service.rb [hunk 1]
```diff
@@ -15,6 +15,8 @@
 # limitations under the License.
 #
 
+require_relative '../helpers/barclamp/neutron_helper.rb'
```
Comment (vuntz) [inline crowbar_framework/app/models/neutron_service.rb]: @tboerger is this something you like, or would you do it another way?

Comment (tboerger) [inline crowbar_framework/app/models/neutron_service.rb]: Helpers don't have to be mixed into models. They've got totally different responsibilities :(

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 2]
```diff
@@ -17,35 +17,27 @@
...
-      selected = selected.gsub(/\s+/, "").split(",")
-      options_for_select(
-        [
-          ["linuxbridge", "linuxbridge"],
-          ["openvswitch", "openvswitch"],
-          ["cisco_nexus", "cisco_nexus"],
-        ],
-        selected
-      )
+    def self.networking_plugins_valid()
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: If there are no parameters remove the ```()``` please. Anyway validation stuff should be placed in the model.

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 3]
```diff
@@ -17,35 +17,27 @@
...
+      ["ml2", "vmware"]
     end
 
-    def networking_ml2_type_drivers_valid()
+    def self.networking_ml2_type_drivers_valid()
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: If there are no parameters remove the ```()``` please. Anyway validation stuff should be placed in the model.

File: crowbar_framework/app/models/neutron_service.rb (hunk 1 above)
Comment (tboerger) [inline crowbar_framework/app/models/neutron_service.rb]: I have commented that validation stuff should be part of the model and not the helper.

Comment (toabctl) [inline crowbar_framework/app/models/neutron_service.rb]: @tboerger I need a list of valid ml2 drivers in the helpers (for the ```options_for_select()``` method) **and** the model (for validation). To which file should I add this list?

Comment (tboerger) [inline crowbar_framework/app/models/neutron_service.rb]: This should be part of the model. And the view (or even the helper which is some layer on top of the view) should access the model methods.

Comment (toabctl) [inline crowbar_framework/app/models/neutron_service.rb]: Done in https://github.com/crowbar/barclamp-neutron/commit/2e545420df705c6ae99da591c1280414647aa26e

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 4]
```diff
@@ -17,35 +17,15 @@
...
-    end
-
-    def networking_ml2_type_drivers_valid()
-      ["vlan", "gre"]
-    end
-
-    def networking_ml2_type_drivers_for_neutron(selected)
-      selected = selected.gsub(/\s+/, "").split(",")
-      valid_options = networking_ml2_type_drivers_valid()
+    def self.options_keep_order(valid_options, selected)
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: ```def options_keep_order(valid_options, selected)```, no need for self.

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 5]
```diff
@@ -58,16 +38,28 @@ def networking_ml2_type_drivers_for_neutron(selected)
...
         options << [el, el]
       end
-      options_for_select(options, selected)
+      options
+    end
+
+    def networking_ml2_mechanism_drivers_for_neutron(selected)
+      selected = selected.gsub(/\s+/, "").split(",")
+      valid_options = NeutronService.networking_ml2_mechanism_drivers_valid
+      options_for_select(NeutronHelper.options_keep_order(valid_options, selected), selected)
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Just call ```options_keep_order``` without the class name in front of it.

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 6]
```diff
@@ -58,16 +38,28 @@ def networking_ml2_type_drivers_for_neutron(selected)
...
+    end
+
+    def networking_ml2_type_drivers_for_neutron(selected)
+      selected = selected.gsub(/\s+/, "").split(",")
+      valid_options = NeutronService.networking_ml2_type_drivers_valid
+      options_for_select(NeutronHelper.options_keep_order(valid_options, selected), selected)
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Just call ```options_keep_order``` without the class name in front of it.

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 7]
```diff
@@ -17,35 +17,15 @@
...
 module Barclamp
   module NeutronHelper
+
     def networking_plugins_for_neutron(selected)
       options_for_select(
-        [
-          ["ml2", "ml2"],
-          ["vmware", "vmware"]
-        ],
+        NeutronService.networking_plugins_valid.map{|el| [el, el]},
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: ```map { |el| [el, el] }``` should be nice :)

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 8]
```diff
@@ -58,16 +38,28 @@ def networking_ml2_type_drivers_for_neutron(selected)
...
+
+    def networking_ml2_type_drivers_for_neutron(selected)
+      selected = selected.gsub(/\s+/, "").split(",")
+      valid_options = NeutronService.networking_ml2_type_drivers_valid
+      options_for_select(NeutronHelper.options_keep_order(valid_options, selected), selected)
     end
 
     def networking_ml2_type_drivers_default_provider_network_for_neutron(selected)
-      options_for_select(networking_ml2_type_drivers_valid().map{|x| [x, x]},
+      options_for_select(NeutronService.networking_ml2_type_drivers_valid().map{|x| [x, x]},
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Skip the empty ```()```

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb [hunk 9]
```diff
@@ -58,16 +38,28 @@ def networking_ml2_type_drivers_for_neutron(selected)
...
                          selected.to_s)
     end
 
     def networking_ml2_type_drivers_default_tenant_network_for_neutron(selected)
-      options_for_select(networking_ml2_type_drivers_valid().map{|x| [x, x]},
+      options_for_select(NeutronService.networking_ml2_type_drivers_valid().map{|x| [x, x]},
```
Comment (tboerger) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Skip the empty ```()```

File: crowbar_framework/app/models/neutron_service.rb [hunk 10]
```diff
@@ -107,6 +119,9 @@ def validate_proposal_after_save proposal
     ml2_type_drivers_default_provider_network = proposal["attributes"]["neutron"]["ml2_type_drivers_default_provider_network"]
     ml2_type_drivers_default_tenant_network = proposal["attributes"]["neutron"]["ml2_type_drivers_default_tenant_network"]
 
+    ml2_type_drivers_valid = NeutronService.networking_ml2_type_drivers_valid
```
Comment (tboerger) [inline crowbar_framework/app/models/neutron_service.rb]: ```self.networking_ml2_type_drivers_valid``` should work, no need for the class name inside of this class.

File: crowbar_framework/app/models/neutron_service.rb [hunk 11]
```diff
@@ -107,6 +119,9 @@ def validate_proposal_after_save proposal
...
+    ml2_mechanism_drivers_valid = NeutronService.networking_ml2_mechanism_drivers_valid
```
Comment (tboerger) [inline crowbar_framework/app/models/neutron_service.rb]: ```self.networking_ml2_type_drivers_valid``` should work, no need for the class name inside of this class.

**Comment by tboerger** [comment]
Just code style comments, otherwise it looks good to me.

File: crowbar_framework/app/models/neutron_service.rb (hunk 11 above)
Comment (toabctl) [inline crowbar_framework/app/models/neutron_service.rb]: No. It doesn't

File: crowbar_framework/app/models/neutron_service.rb (hunk 10 above)
Comment (toabctl) [inline crowbar_framework/app/models/neutron_service.rb]: No. It doesn't

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb (hunk 9 above)
Comment (toabctl) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Done in https://github.com/crowbar/barclamp-neutron/commit/ed72a9a35db400e62101fa774ed32947ee5880b1

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb (hunk 8 above)
Comment (toabctl) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Done in https://github.com/crowbar/barclamp-neutron/commit/ed72a9a35db400e62101fa774ed32947ee5880b1

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb (hunk 7 above)
Comment (toabctl) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Done in https://github.com/crowbar/barclamp-neutron/commit/ed72a9a35db400e62101fa774ed32947ee5880b1

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb (hunk 6 above)
Comment (toabctl) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Done in https://github.com/crowbar/barclamp-neutron/commit/ed72a9a35db400e62101fa774ed32947ee5880b1

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb (hunk 5 above)
Comment (toabctl) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Done in https://github.com/crowbar/barclamp-neutron/commit/ed72a9a35db400e62101fa774ed32947ee5880b1

File: crowbar_framework/app/helpers/barclamp/neutron_helper.rb (hunk 4 above)
Comment (toabctl) [inline crowbar_framework/app/helpers/barclamp/neutron_helper.rb]: Done in https://github.com/crowbar/barclamp-neutron/commit/ed72a9a35db400e62101fa774ed32947ee5880b1

File: crowbar_framework/app/models/neutron_service.rb (hunk 11 above)
Comment (tboerger) [inline crowbar_framework/app/models/neutron_service.rb]: Ah yeah, you are right... ```self.class``` :D

File: crowbar_framework/app/models/neutron_service.rb (hunk 10 above)
Comment (tboerger) [inline crowbar_framework/app/models/neutron_service.rb]: Ah yeah, you are right... ```self.class.method_name``` :D

**Comment by vuntz** [comment]
+1

**Pull request merged**